### PR TITLE
Merge master 2.9.99

### PR DESCRIPTION
**Pull request by ndellingwood**

### CODE REVIEW

**Comment by ndellingwood** [comment]
Master branch updated with kokkos 2.9.99, this PR merges master back into develop.

File: BUILD.md [hunk 1]
```diff
@@ -157,9 +157,6 @@ Options can be enabled by specifiying `-DKokkos_ENABLE_X`.
 * Kokkos_ENABLE_DEPRECATED_CODE
     * Whether to enable deprecated code
     * BOOL Default: OFF
-* Kokkos_ENABLE_EXAMPLES
-    * Whether to enable building examples
-    * BOOL Default: OFF
```
Comment (masterleinad) [inline BUILD.md]: We don't want this here.

File: CMakeLists.txt [hunk 2]
```diff
@@ -120,14 +120,18 @@ ENDIF()
...
+GLOBAL_RESET(KOKKOS_CUDAFE_OPTIONS)
+GLOBAL_RESET(KOKKOS_XCOMPILER_OPTIONS)
 # We need to append text here for making sure TPLs
 # we import are available for an installed Kokkos
-GLOBAL_SET(KOKKOS_TPL_EXPORTS)
+GLOBAL_RESET(KOKKOS_TPL_EXPORTS)
+# We need these for controlling the exact -std flag
+GLOBAL_RESET(KOKKOS_DONT_ALLOW_EXTENSIONS)
+GLOBAL_RESET(KOKKOS_USE_CXX_EXTENSIONS)
+GLOBAL_RESET(KOKKOS_CXX_STANDARD_FEATURE)
```
Comment (masterleinad) [inline CMakeLists.txt]: Decided against introducing `GLOBAL_RESET`.

File: cm_generate_makefile.bash [hunk 3]
```diff
@@ -343,5 +341,5 @@ if [[ ${COMPILER} == *clang* ]]; then
    fi 
 fi
  
-echo cmake $COMPILER_CMD  -DCMAKE_CXX_FLAGS="${KOKKOS_CXXFLAGS}" -DCMAKE_EXE_LINKER_FLAGS="${KOKKOS_LDFLAGS}" -DCMAKE_INSTALL_PREFIX=${PREFIX} ${KOKKOS_DEVICE_CMD} ${KOKKOS_ARCH_CMD} -DKokkos_ENABLE_TESTS=ON -DKokkos_ENABLE_EXAMPLES=${KOKKOS_DO_EXAMPLES} ${KOKKOS_OPTION_CMD} ${KOKKOS_CUDA_OPTION_CMD} -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_CXX_EXTENSIONS=OFF ${STANDARD_CMD} ${KOKKOS_DEBUG_CMD} ${KOKKOS_PATH}
```
Comment (masterleinad) [inline cm_generate_makefile.bash]: This should stay.

File: cmake/fake_tribits.cmake [hunk 4]
```diff
@@ -24,6 +24,10 @@ IF(NOT ${MACRO_DEFINE_NAME} STREQUAL "")
 ENDIF()
 ENDMACRO()
 
+MACRO(GLOBAL_RESET VARNAME)
+  SET(${VARNAME} "" CACHE INTERNAL "" FORCE)
+ENDMACRO()
```
Comment (masterleinad) [inline cmake/fake_tribits.cmake]: and this

File: cmake/kokkos_enable_options.cmake [hunk 5]
```diff
@@ -30,7 +30,6 @@ KOKKOS_ENABLE_OPTION(CUDA_UVM             OFF "Whether to use unified memory (UM
 KOKKOS_ENABLE_OPTION(CUDA_LDG_INTRINSIC   OFF "Whether to use CUDA LDG intrinsics")
 KOKKOS_ENABLE_OPTION(HPX_ASYNC_DISPATCH   OFF "Whether HPX supports asynchronous dispatch")
 KOKKOS_ENABLE_OPTION(TESTS         OFF  "Whether to build the unit tests")
-KOKKOS_ENABLE_OPTION(EXAMPLES      OFF  "Whether to build the examples")
```
Comment (masterleinad) [inline cmake/kokkos_enable_options.cmake]: and this

File: cmake/kokkos_tribits.cmake [hunk 6]
```diff
@@ -393,15 +391,3 @@ MACRO(KOKKOS_ADD_TEST_DIRECTORIES)
...
-  if (KOKKOS_HAS_TRILINOS)
-    TRIBITS_ADD_EXAMPLE_DIRECTORIES(${ARGN})
-  else()
-    IF(KOKKOS_ENABLE_EXAMPLES)
-      FOREACH(EXAMPLE_DIR ${ARGN})
-        ADD_SUBDIRECTORY(${EXAMPLE_DIR})
-      ENDFOREACH()
-    ENDIF()
-  endif()
-ENDMACRO()
```
Comment (masterleinad) [inline cmake/kokkos_tribits.cmake]: and this

File: example/CMakeLists.txt [hunk 7]
```diff
@@ -2,19 +2,19 @@
...
+#KOKKOS_ADD_EXAMPLE_DIRECTORIES(fenl)
+#KOKKOS_ADD_EXAMPLE_DIRECTORIES(multi_fem)
+#KOKKOS_ADD_EXAMPLE_DIRECTORIES(md_skeleton)
+#KOKKOS_ADD_EXAMPLE_DIRECTORIES(global_2_local_ids)
+#KOKKOS_ADD_EXAMPLE_DIRECTORIES(grow_array)
+#KOKKOS_ADD_EXAMPLE_DIRECTORIES(sort_array)
+#if(NOT Kokkos_ENABLE_CUDA)
+#KOKKOS_ADD_EXAMPLE_DIRECTORIES(tutorial)
+#endif()
+#KOKKOS_SUBPACKAGE_POSTPROCESS()
```
Comment (masterleinad) [inline example/CMakeLists.txt]: This should be resolved in the other pull request (#2688).

File: generate_makefile.bash [hunk 8]
```diff
@@ -81,9 +79,6 @@ do
       echo "Warning: ${key} is deprecated"
       echo "Call make with appropriate -j flag"
       ;;
-    --no-examples)
```
Comment (masterleinad) [inline generate_makefile.bash]: This file should likely stay the same.

File: scripts/testing_scripts/generate_makefile.bash [hunk 9]
```diff
@@ -2,8 +2,6 @@
 
 KOKKOS_DEVICES=""
 
-KOKKOS_DO_EXAMPLES="1"
```
Comment (masterleinad) [inline scripts/testing_scripts/generate_makefile.bash]: and this

File: master_history.txt [hunk 10]
```diff
@@ -17,3 +17,4 @@ tag:  2.7.00     date: 05:24:2018    master: e01945d0    develop: 2d13f608
 tag:  2.7.24     date: 11:04:2018    master: d3a94192    develop: 7a06fc81
 tag:  2.8.00     date: 02:05:2019    master: 34931a36    develop: d1659d1d
 tag:  2.9.00     date: 06:24:2019    master: 5d6e7fb3    develop: 4c6cb80a
+tag:  2.9.99     date: 01:31:2020    master: 2983b80d    release-candidate-3.0: fdc904a6
```
Comment (masterleinad) [inline master_history.txt]: We need this.

File: cmake/kokkos_tribits.cmake [hunk 11]
```diff
@@ -43,8 +43,6 @@ MACRO(KOKKOS_SUBPACKAGE NAME)
     SET(PACKAGE_NAME ${PACKAGE_NAME}${NAME})
     STRING(TOUPPER ${PACKAGE_NAME} PACKAGE_NAME_UC)
     SET(${PACKAGE_NAME}_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
-    #ADD_INTERFACE_LIBRARY(PACKAGE_${PACKAGE_NAME})
-    #GLOBAL_SET(${PACKAGE_NAME}_LIBS "")
```
Comment (masterleinad) [inline cmake/kokkos_tribits.cmake]: This looks reasonable.

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,5 +1,45 @@
...
+- OffsetView: Kokkos::OffsetView: allow offset=0 [\#2246](https://github.com/kokkos/kokkos/issues/2246)
+- DeepCopy: Missing DeepCopy instrumentation in Kokkos [\#2522](https://github.com/kokkos/kokkos/issues/2522)
+- nvcc\_wrapper: --host-only fails with mutiple -W\* flags [\#2484](https://github.com/kokkos/kokkos/issues/2484)
+- nvcc\_wrapper: taking first -std option is counterintuitive [\#2553](https://github.com/kokkos/kokkos/issues/2553)
+- Subview: Error taking subviews of views with static_extents of min rank [\#2448](https://github.com/kokkos/kokkos/issues/2448)
+- TeamPolicy: reducers with valuetypes without += broken on CUDA [\#2410](https://github.com/kokkos/kokkos/issues/2410)
+- Libs: Fix inconsistency of Kokkos library names in Kokkos and Trilinos [\#1902](https://github.com/kokkos/kokkos/issues/1902)
+- Complex: operator\>\> for complex\<T\> uses std::ostream, not std::istream [\#2313](https://github.com/kokkos/kokkos/issues/2313)
+- Macros: Restrict not honored for non-intel compilers  [\#1922](https://github.com/kokkos/kokkos/issues/1922)
```
Comment (masterleinad) [inline CHANGELOG.md]: We need this here.

**Comment by dalg24** [comment]
> Master branch updated with kokkos 2.9.99, this PR merges master back into develop.

Why would we merge master into develop?

**Comment by stanmoore1** [comment]
See the comment here: https://github.com/kokkos/kokkos/pull/1223

**Comment by masterleinad** [comment]
Hmm... if we really want to do this, let's first merge #2688 and then rebase/remerge this one. Hopefully, this results in far fewer changes. After that, we might want to cherry-pick #2709 here.

I think that we talked about making the development branch (independent of that being called `master` or `develop`) the default branch in the long run such that we don't need to do this in the future.

**Comment by ndellingwood** [comment]
> let's first merge #2688 and then rebase/remerge this one

I'm in favor of this, I can close this PR and reissue once that is merged in.

**Comment by ndellingwood** [comment]
Closing in favor of #2727

**Pull request closed without merging**